imapserver/imapmemserver/mailbox.go

```diff
@@ -418,13 +431,40 @@ func (mbox *MailboxView) staticSearchCriteria(criteria *imap.SearchCriteria) {
 }
 
 func (mbox *MailboxView) Store(w *imapserver.FetchWriter, numSet imap.NumSet, flags *imap.StoreFlags, options *imap.StoreOptions) error {
-	mbox.forEach(numSet, func(seqNum uint32, msg *message) {
-		msg.store(flags)
-		mbox.Mailbox.tracker.QueueMessageFlags(seqNum, msg.uid, msg.flagList(), mbox.tracker)
+	var modifiedSeqNums []uint32
+	var modifiedMsgs []*message
+
+	mbox.mutex.Lock()
+	mbox.forEachLocked(numSet, func(seqNum uint32, msg *message) {
+		if options != nil && options.UnchangedSince > 0 && msg.modSeq > options.UnchangedSince {
+			return
+		}
+
+		if changed := msg.store(flags); changed {
+			mbox.highestModSeq++
+			msg.modSeq = mbox.highestModSeq
+
+			modifiedSeqNums = append(modifiedSeqNums, seqNum)
+			modifiedMsgs = append(modifiedMsgs, msg)
+		}
 	})
-	if !flags.Silent {
-		return mbox.Fetch(w, numSet, &imap.FetchOptions{Flags: true})
+	mbox.mutex.Unlock()
+
+	for i, seqNum := range modifiedSeqNums {
+		msg := modifiedMsgs[i]
+		mbox.Mailbox.tracker.QueueMessageFlags(seqNum, msg.uid, msg.flagList(), mbox.tracker)
```

Comment:
Is there a reason to move the `QueueMessageFlags` call outside of the `mbox.forEach` loop?

`msg.flagList` can only be called with the mailbox mutex locked. `QueueMessageFlags` never blocks.